moxin-protocol/src/protocol.rs

```diff
@@ -1,21 +1,142 @@
-use crate::data::Model;
+use crate::data::*;
+use std::sync::mpsc::Sender;
+use chrono::NaiveDate;
+use anyhow::Result;
 
 #[derive(Clone, Debug)]
-pub enum Command {
-    GetFeaturedModels,
+pub enum FileDownloadResponse {
+    Progress(FileID, f32),
+    Completed(DownloadedFile),
+}
 
-    // The argument is a string with the keywords to search for.
-    SearchModels(String),
+#[derive(Clone, Debug)]
+pub enum ContextOverflowPolicy {
+    StopAtLimit,
+    TruncateMiddle,
+    TruncatePastMessages,
+}
+
+#[derive(Clone, Debug)]
+pub enum GPULayers {
+    Specific(u32),
+    Max,
+}
+
+#[derive(Clone, Debug)]
+pub struct LoadModelOptions {
+    pub prompt_template: Option<String>,
+    pub gpu_layers: GPULayers,
+    pub use_mlock: bool,
+    pub n_batch: u32,
+    pub n_ctx: u32,
+    pub rope_freq_scale: f32,
+    pub rope_freq_base: f32,
 
-    // The argument is the File name.
-    DownloadFile(String),
+    // TBD Not really sure if this is something backend manages or if it is matter of
+    // the client (if it is done by tweaking the JSON payload for the chat completition)
+    pub context_overflow_policy: ContextOverflowPolicy
 }
 
 #[derive(Clone, Debug)]
-pub enum Response {
-    // Response to the GetFeaturedModels command
-    FeaturedModels(Vec<Model>),
+pub struct LoadedModelInfo {
+    pub file_id: FileID,
+    pub model_id: ModelID,
+
+    // JSON formatted string with the model information. See "Model Inspector" in LMStudio.
+    pub information: String,
+}
+
+#[derive(Clone, Debug)]
+pub struct ModelResourcesInfo {
+    ram_usage: f32,
+    cpu_usage: f32,
+}
+
+#[derive(Clone, Debug)]
+pub enum LoadModelResponse {
+    Progress(FileID, f32),
+    Completed(LoadedModelInfo),
+    ModelResoucesUsage(ModelResourcesInfo),
+}
+
+pub enum ChatError {
+    BackendNotRun,
+    EndOfSequence,
+    ContextFull,
+    PromptTooLong,
+    TooLarge,
+    InvalidEncoding,
+    Other,
+}
+
+#[derive(Clone, Debug)]
+pub enum StopReason {
+    Completed,
+    Stopped
+}
+
+#[derive(Clone, Debug)]
+pub struct ChatCompletionData {
+    // The response from the model in JSON format
+    // https://platform.openai.com/docs/api-reference/chat/object
+    response: String,
+
+    // The remaining fields are stats about the chat completion process
+    time_to_first_token: f32,
+    time_to_generate: f32,
+    speed: f32,
+    gpu_layers: u32,
+    cpu_threads: u32,
+    mlock: bool,
+    token_count: u32,
+    token_limit: u32,
```

Comment:
Why is it necessary to include this metadata in ChatCompletion?

Reply:
This is a feature in LMStudio:

<img width="1049" alt="image" src="https://github.com/project-robius/moxin/assets/487140/06b3fc69-93ea-481f-8d42-51173f4ff6a1">

Not saying we have to implement everything at once! However, we are trying to capture here everything we would need to mimic LMStudio features. For instance, I would say that it is OK if the initial versions of the backend implementation does not care about those fields (leaving it in zero or default value).

Reply:
Ok, I see. My understanding is that `gpu_layers`, `cpu_threads`, and `mlock` come from the frontend. The frontend is aware of them and their changes.

Reply:
> Ok, I see. My understanding is that gpu_layers, cpu_threads, and mlock come from the frontend. The frontend is aware of them and their changes.

Good point.. so I can simply use the values we sent in the first place. Makes sense 👍